### ISSUE TITLE
Pre-compilation of styles for production

### ISSUE THREAD

**Issue by monolithed** [ant-design/ant-design#44770]
### What problem does this feature solve?

When I was trying to integrate styles, I encountered a problem where the server was very slow in delivering the page. I measured the Time To First Byte (TTFB) and observed that the server was taking approximately 2-3 seconds to fetch the styles at runtime, depending on the computational power.

My first thought was to drop @ant-design/cssinjs and build the styles without it. However, this attempt was unsuccessful because of FOUC (Flash of Unstyled Content).

<br/>

https://github.com/ant-design/ant-design/assets/803674/a5228765-d3fc-4c63-a8df-4b6da000b01d

<br/>
<br/>

**Here's my code:**

```ts
import fs from 'node:fs';
import {extractStyle} from '@ant-design/static-style-extract';

fs.writeFileSync('./public/antd.min.css', extractStyle()); // NetxJS's public directory
```

```tsx
import React from 'react';

const RootLayout: React.FunctionComponent<React.PropsWithChildren> = ({children}) => (
    <html lang="en">
        <head>
            <link rel="stylesheet" type="text/css" href="/antd.min.css" />
        </head>

        <body>
            {children}
        </body>
    </html>
);

export default RootLayout;
```

```tsx
import {App, Button, ConfigProvider} from 'antd';

export default function Home () {
    return (
        <App>
            <ConfigProvider theme={{
                components: {
                    Button: {  // the styles aren't applied due to this line
                        colorText: 'red',
                    }
                }
            }}>
                <Button>Text</Button>
            </ConfigProvider>
        </App>
    )
}
```

### What does the proposed API look like?

It'd be nice to have only one copy of the styles that can be reused as a separate file hosted on a CDN without `@ant-design/cssinjs` (like v4).

**Comment by afc163** [comment]
Could you provide a reproduce repository?

**Comment by MadCcc** [comment]
More of a problem with Nextjs not antd. Dynamicly write css file with SSR is not recommended, and `/public` of Nextjs only holds static files.

**Comment by MadCcc** [comment]
Ref: https://github.com/ant-design/ant-design-examples/blob/main/examples/with-nextjs-generate-css-on-demand/pages/_document.tsx

**Comment by monolithed** [comment]
@afc163, it turns out, the issue is with ConfigProvider! If you wrap the component in it with property override, for example, components.Button, the styles are not applied.

```tsx
import {Button, ConfigProvider} from 'antd';
import styles from './page.module.css';

export default function Home () {
    return (
        <main>
            <ConfigProvider theme={{
                components: {
                    Button: {  // this line
                        colorText: 'red',
                    }
                }
            }}>
                <Button>Text</Button>
            </ConfigProvider>
        </main>
    )
}
```

And the second issue is that the styles are loaded twice: 
1) They are always included in the Ant Design bundle
2) An additional 1mb of CSS is loaded through the link tag.

**Comment by monolithed** [comment]
> Ref: https://github.com/ant-design/ant-design-examples/blob/main/examples/with-nextjs-generate-css-on-demand/pages/_document.tsx

I've tried this approach, but it's intended to be executed at runtime, meaning multiple CSS files will be generated for each user. This puts a load on the server, and users won't see the page for at least 2.7-3 seconds. Additionally, dynamically creating files can pose security risks, and services like Vercel prohibit such practices.

**Comment by monolithed** [comment]
@afc163, @MadCcc here's a reproducible repo https://github.com/Papacardo/test

**Comment by MadCcc** [comment]
Did you expected a css file like what 4.x provided?
Since styles are dynamic with different theme configuration, it is recommended to extract it with `extractStyle` of `@ant-design/cssinjs`.
Is that enough or any other problems?

**Comment by monolithed** [comment]
> Did you expected a css file like what 4.x provided?
> Since styles are dynamic with different theme configuration, it is recommended to extract it with `extractStyle` of `@ant-design/cssinjs`.
> Is that enough or any other problems?

@MadCcc, the main problem is that it's impossible to compile the styles into a separate file to reuse it. 
The suggestion to use `@ant-design/cssinjs` with style extraction seems pointless because the styles would be duplicated in three places: the first place is the `antd`,  where styles are is always included in the library and cannot be excluded from there; the second one is the bundle with extracted styles, and the third is the client library `@ant-design/cssinjs`, which also contains a bunch of code and the same styles. 
In other words, the method of style extraction combined with `@ant-design/cssinjs` generates more styles than the entire Ant Design library itself.

It would be desirable to have a way to compile styles into a single bundle without duplicating them in three places.

**Comment by monolithed** [comment]
<img width="692" alt="Screenshot 2023-09-15 at 15 50 47" src="https://github.com/ant-design/ant-design/assets/803674/90e271c8-fd25-4087-a9c1-55cea6063f08">

**Comment by monolithed** [comment]
@afc163, @MadCcc, @zombieJ, @kiner-tang, to summarize, the seriousness of the situation lies in the fact that there is no production-ready solution available. In all cases, SSR is required, and there are always duplicates of builds:

* [with-nextjs-inline-style](https://github.com/ant-design/ant-design-examples/tree/main/examples/with-nextjs-inline-style) — styles always come with a FOUC (Flash of Unstyled Content) effect. **It is not suitable for production**.
* [with-nextjs-generate-css-on-demand](https://github.com/ant-design/ant-design-examples/tree/main/examples/with-nextjs-generate-css-on-demand) — a similar solution to `with-nextjs-app-router-inline-style`, but additionally creates copies of style files for each user on the server. It is potentially unsafe, puts a heavy load on the server, and results in a prolonged initial loading time. **It is not suitable for production**.
* [with-nextjs-extract-style](https://github.com/ant-design/ant-design-examples/tree/main/examples/with-nextjs-extract-style) — it required `@ant-design/cssinjs` (without this dependency, ConfigProvider cannot detect custom styles). This implies that with SSR enabled, styles will be compiled at runtime, taking 3 seconds. It's a method similar to `with-nextjs-app-router-inline-style` and **is not suitable for production**.
* [with-nextjs-app-router-inline-style](https://github.com/ant-design/ant-design-examples/tree/main/examples/with-nextjs-app-router-inline-style) — styles are compiled at runtime with an average build speed of 3 seconds. This method is convenient only for development and **is not suitable for production**.

If any of you knows a way to connect styles without duplicating them and without SSR, please let me know.

**Comment by monolithed** [comment]
I would like to understand why the ticket is closed. Currently, it is impossible to extract styles into a separate file. All the suggested methods are not suitable for production because they require runtime compilation.

**Comment by lukaszczernal** [comment]
@monolithed did a really nice job creating a summary of mentioned issue. His ticket should not be cloesed. 
There is a serious problem with using AntD v5 with SSR right now (NextJS). A a possibility to extract CSS to a file during build time would also be future-proof solution. 

There is a lot of movement in how we write CSS lately (css-in-js, functional classes, css modules...) and because of that my company decided to move from css-in-js and css-modules and css-variables. We are in a long run so we need a stable solution.
